### DISCUSSION TITLE
How to: teal modules in embeded environment

### DISCUSSION THREAD

**Discussion by barsoosayque** [teal-language/tl#635] [Q&A]
I'm trying to embed teal compiler into a rust project to generate lua code in runtime. My current generator is:
```lua
-- <...> code to load tl

local env = tl.init_env(false, true, '5.1', nil)
return tl.gen([[
    <teal code>
]], env)
```
I use luajit to run this generator when I'm loading tl files. Isolated scripts are ok, but when script has `require("local_module")`, I don't see a way to include this module so `tl.gen` would actually produce code (it just throws an error about an unknown module). Using standalone tl tool to check my scripts return no error, since I have a valid `tlconfig.lua` file with `include_dir` pointing to my modules.

So my question is: in an environment when I have a chunk of teal code and a bunch of modules from `include_dir` in form of teal code, how do I compile this chunk into a lua code ?

(I tried setting `env.modules["mymodule"] = [[ <teal module code> ]]`, it does help with resolving a module, but it doesn't provide global records I defined in that module. My current workaround is to regex replace `require`s with actual module code, but that's a very limited approach and it make diagnostic messages misleading, since the line number becomes offset)

**Comment by lenscas** [comment]
there should be an `tl.init_env` function that should help you properly initialize the enviroment that you have to pass to `tl.gen`

I am doing something similar in https://github.com/lenscas/tealr_doc_gen/blob/f9ce06346b0e97a836d36b86f908e5b7bb229953/base_template.etlua#L137

**Reply by barsoosayque** [reply]
Yep, I already use it (forgot to use `env` in a code snippet, fixed now). It helps with generating Lua 5.1 compatible code (thought I didn't test it thoroughly), but I don't see what I can do with it to provide modules for `require` resolution. Maybe I misused the term 'environment' here, I was referencing to the 'project environment' in which there are libraries in `include_dir` and sources in `source_dir`. Basically, how do I put my `tlconfig.lua` into `tl.gen` when I don't have access to lua's os/io module ?

Here is what my test script is:
```lua
-- test.tl
require("quest")

local q = Quest:new()
q.title_id = "test-quest-title"

return q
```
And the library is:
```lua
-- lib/quest.d.tl
global record Quest
    title_id: string

    new: function(): Quest
end
```
When I compile `test.tl`, I get "unknown module 'quest'". And that's what I'm trying to understand, how do I provide `quest.d.tl` to `tl.gen`

**Reply by hishamhm** [reply]
@barsoosayque the `env` accumulates the info on required files. If you `tl.load` the quest.d.tl file into the `env` and then reuse that same `env` when running `tl.gen` on the script, that should allow it to resolve the require.

**Reply by barsoosayque** [reply]
Can you please elaborate on this ?

I tried this and got confused: how do I actually put modules in the `env` ? `tl.load` syntax/type-checks and then return compiled code. To my understanding, I then have to put this compiled module code into `env.loaded`/`env.modules`, but since it's a lua code, it wouldn't provide any types and then my `tl.gen` would fail due to unknown type (and even earlier, because I use modules in other modules).

**Reply by hishamhm** [reply]
passing an `env` to `tl.load` modifies that `env` — it should put it into `env.loaded`. Doesn't it? (I might be mistaken if this is being done by `require` instead...)
